server/map_tree_updater.go

```diff
@@ -168,20 +164,49 @@ func (t *mapTreeUpdater) newTXFunc(tx storage.MapTreeTX) txFunc {
 	}
 }
 
-// doPreload causes the subtreeCache in tx to become populated with all
-// subtrees on the Merkle path for the indices specified in upd.
-// This is a performance workaround for locking issues which occur when the
-// sparse Merkle tree code is used with a single transaction (and therefore a
-// single subtreeCache too).
-func doPreload(ctx context.Context, tx storage.MapTreeTX, depth uint, upd []smt.Node, rev int64) (map[tree.NodeID2][]byte, error) {
+// doPreload loads all Merkle tree tiles that will be involved when updating
+// the given set of nodes at the specified revision.
+func (t *mapTreeUpdater) doPreload(ctx context.Context, tx storage.MapTreeTX, nodes []smt.Node, rev int64) (*smt.TileSet, error) {
 	ctx, spanEnd := spanFor(ctx, "doPreload")
 	defer spanEnd()
-
 	// TODO(pavelkalinnikov): Avoid using HStar3 directly.
-	hs, err := smt.NewHStar3(upd, nil, depth, 0)
+	hs, err := smt.NewHStar3(nodes, nil, uint(t.hasher.BitLen()), 0)
```

Comment:
I like this idea. There are a bunch of places throughout the code where this could be done, so let me put a TODO for now and address it in a follow-up.